vshard/router/init.lua

```diff
@@ -1741,7 +1756,13 @@ end
 local function legacy_cfg(cfg)
     if M.static_router then
         -- Reconfigure.
-        router_cfg(M.static_router, cfg, false)
+        router_cfg_fiber_safe(M.static_router, cfg, false)
+    elseif M.routers[STATIC_ROUTER_NAME] then
+        -- M.static_router is still not set but M.routers[STATIC_NAME] is
+        -- already there. This can only mean, that configuration is still in
+        -- progress. But simultaneous configuration is prohibited.
+        error(lerror.vshard(lerror.code.ROUTER_CFG_IS_IN_PROGRESS,
+                            STATIC_ROUTER_NAME))
     else
         -- Create new static instance.
         local router, err = router_new(STATIC_ROUTER_NAME, cfg)
```

Comment:
Wouldn't this work too?
```Diff
@@ -1754,16 +1754,8 @@ end
 -- static `vshard.router.cfg()` API.
 --
 local function legacy_cfg(cfg)
-    if M.static_router then
-        -- Reconfigure.
-        router_cfg_fiber_safe(M.static_router, cfg, false)
-    elseif M.routers[STATIC_ROUTER_NAME] then
-        -- M.static_router is still not set but M.routers[STATIC_NAME] is
-        -- already there. This can only mean, that configuration is still in
-        -- progress. But simultaneous configuration is prohibited.
-        error(lerror.vshard(lerror.code.ROUTER_CFG_IS_IN_PROGRESS,
-                            STATIC_ROUTER_NAME))
-    else
+    local router = M.routers[STATIC_ROUTER_NAME]
+    if not router then
         -- Create new static instance.
         local router, err = router_new(STATIC_ROUTER_NAME, cfg)
         if router then
@@ -1773,6 +1765,9 @@ local function legacy_cfg(cfg)
         else
             return nil, err
         end
+    else
+        -- Reconfigure.
+        router_cfg_fiber_safe(router, cfg, false)
     end
 end
```

Reply:
Great solution